store/store.go

```diff
@@ -245,82 +275,181 @@ func (bs *BlockStore) LoadSeenCommit(height int64) *types.Commit {
 	return commit
 }
 
-// PruneBlocks removes block up to (but not including) a height. It returns number of blocks pruned.
+// PruneBlocks removes block up to (but not including) a height. It returns the number of blocks pruned.
 func (bs *BlockStore) PruneBlocks(height int64) (uint64, error) {
 	if height <= 0 {
 		return 0, fmt.Errorf("height must be greater than 0")
 	}
-	bs.mtx.RLock()
-	if height > bs.height {
-		bs.mtx.RUnlock()
-		return 0, fmt.Errorf("cannot prune beyond the latest height %v", bs.height)
+
+	if height > bs.Height() {
+		return 0, fmt.Errorf("height must be equal to or less than the latest height %d", bs.Height())
+	}
+
+	// remove block meta first as this is used to indicate whether the block exists.
+	// For this reason, we also use ony block meta as a measure of the amount of blocks pruned
+	pruned, err := bs.pruneBlockMetaAndHashes(height)
+	if err != nil {
+		return pruned, err
 	}
-	base := bs.base
-	bs.mtx.RUnlock()
-	if height < base {
-		return 0, fmt.Errorf("cannot prune to height %v, it is lower than base height %v",
-			height, base)
+
+	if err := bs.batchDelete(blockPartKey(1, 0), blockPartKey(height, 0)); err != nil {
+		return pruned, err
 	}
 
-	pruned := uint64(0)
+	if err := bs.batchDelete(blockCommitKey(1), blockCommitKey(height)); err != nil {
+		return pruned, err
+	}
+
+	if err := bs.batchDelete(seenCommitKey(1), seenCommitKey(height)); err != nil {
+		return pruned, err
+	}
+
+	return pruned, nil
+}
+
+// pruneBlockMetaAndHashes prunes all block metas from the base height to the retainHeight
+// but before each, loads the hash from the block meta to also prune from the block hash table
+func (bs *BlockStore) pruneBlockMetaAndHashes(retainHeight int64) (uint64, error) {
```

Comment:
It's a bit sad that we have to duplicate all of this. If `batchDelete` took a callback function `func(key, value []byte, batch dbm.Batch) error` then we could pass `nil` in the typical case, and a function that extracts the block meta and adds the hash delete to the batch for the block case.

Reply:
Okay I can try add something like this